### ISSUE TITLE
string-format does not support external modules

### ISSUE THREAD

**Issue by matthew-nichols** [mgechev/revive#1061]
**Is your feature request related to a problem? Please describe.**
We currently use an (in-house) logging library (e.g., `github.com/example/common/logger`, type `Logger`, member functions `Info`, `Warn`, `Crit`, etc) that our microservices import, and we would like to enforce that logging messages do not start with lowercase characters.

**Describe the solution you'd like**
For `string-format` to be able to reference functions/types outside the current module and the standard library.

**Comment by chavacava** [comment]
Hi @matthew-nichols, apologies for the late response. 
The [configuration options of the rule](https://github.com/mgechev/revive/blob/master/RULES_DESCRIPTIONS.md#string-format) seem to allow addressing your use case. Can you please confirm?

**Comment by matthew-nichols** [comment]
Not by my reading of it, and I read that before submitting this ticket.

There doesn't seem to be a way to say, for instance, `"github.com/op/go-logging".Logger.Criticalf` (I know I've also tried the equivalent of `logging.Logger.Criticalf` and it didn't work and a couple other variants I don't remember, none of them worked).

(We don't use this exact library https://pkg.go.dev/github.com/op/go-logging#Logger.Criticalf, but it seems close enough to what this rule is missing)

**Comment by chavacava** [comment]
Analyzing the following code

```go
package main

import mylogger "github.com/example/common/logger"

func main() {
	mylogger.Info("Starting...")
	mylogger.Warn("Doing nothing")
	mylogger.Crit("Stop wasting CPU")
}
```
with the following configuration
```toml
[rule.string-format]
  arguments = [
    ["mylogger.Info[0],mylogger.Warn[0],mylogger.Crit[0]", "/^([^A-Z]|$)/", "must not start with a capital letter"],
    ]
```
works as expected. The same for

```go
package main

import "github.com/example/common/logger"

func main() {
	logger.Info("Starting...")
	logger.Warn("Doing nothing")
	logger.Crit("Stop wasting CPU resources")
}
```
and
```toml
[rule.string-format]
  arguments = [
    ["logger.Info[0],logger.Warn[0],logger.Crit[0]", "/^([^A-Z]|$)/", "must not start with a capital letter"],
  ]
```

Do these examples model well your use case?

**Comment by matthew-nichols** [comment]
FYI it needs to be a member function of a type in the library, so no, that does not, needs to be more like

```go
package main

import "github.com/example/common/logger"

func main() {
        log := logger.New("status", "in-github-issue")
	log.Info("Starting...")
	log.Warn("Doing nothing")
	log.Crit("Stop wasting CPU resources")
}
```

**Comment by chavacava** [comment]
Analyzing your example (`deleteme.go`)

```go
package main

import "github.com/example/common/logger"

func main() {
        log := logger.New("status", "in-github-issue")
	log.Info("Starting...")
	log.Warn("Doing nothing")
	log.Crit("Stop wasting CPU resources")
}
```

with

```toml
[rule.string-format]
  arguments = [
    ["log.Info[0],log.Warn[0],log.Crit[0]", "/^([^A-Z]|$)/", "must not start with a capital letter"],
  ]
```
results in

```
deletememe.go:7:11: must not start with a capital letter
deletememe.go:8:11: must not start with a capital letter
deletememe.go:9:11: must not start with a capital letter
```

**Comment by matthew-nichols** [comment]
So does that mean the logging variable name needs to be consistent? Unless there's a lint to do that (and I don't even think that's possible, we often have to rename it inside for loops for instance, unless we can match a prefix like `"clog.*\.Info[0]"`) I'm not too sure how well this would work.

This does not work at all how I would think it would... is it not possible for string-format to reference actual types?

**Comment by chavacava** [comment]
> is it not possible for string-format to reference actual types?

The rule implementation does not leverage type information. The  rule, thus its configuration, is not type-aware but function-name-aware.

>So does that mean the logging variable name needs to be consistent?

Yes. Keeping logging variables consistent will allow simpler linter configurations. Besides what concerns to the linter rule itself, keeping consistent naming for things like the logger is a good practice in general (I currently work on a +600k lines code base where the logger instance is `LOGGER` everywhere) You can also [add your own linting rule](https://github.com/mgechev/revive?tab=readme-ov-file#using-revive-as-a-library) to enforce the naming of the logger instance.

**Comment by matthew-nichols** [comment]
> The rule implementation does not leverage type information. The rule, thus its configuration, is not type-aware but function-name-aware.

I feel like the documentation for string-format needs this added, possibly in bold, because as written it sounds like it matches on types, not on whatever name it happened to show up as (and is also just extremely surprising).

Also would it match/not match in situations like this?

```go

type t struct {
  clog *logger
  ... etc
}

func (this *t) Run() {
  t.clog.Warn("run started")
}

func main() {
  (&t{clog: logger.New("func", "main")}).Run()
}
```

This is also an extremely common (though simplified) pattern in our codebase.

**Comment by chavacava** [comment]
> I feel like the documentation for string-format needs this added

I'll try to make the doc more explicit on that.

About the last pattern, again, if there is some convention on names, a simple configuration does the job:

(`deleteme.go`)
```go
package main

type t struct {
  clog *logger.Logger
}

func (this *t) Run() {
  t.clog.Warn("Run started")
}

func main() {
  (&t{clog: logger.New("func", "main")}).Run()
}
```
with 
```toml
[rule.string-format]
  arguments = [
    ["clog.Warn[0]", "/^([^A-Z]|$)/", "must not start with a capital letter"],
  ]
```
produces
```
deletememe.go:8:15: must not start with a capital letter
```

I'll close the issue for now. If there are new requests on supporting types we can re-open the discussion.
